lib/developer_portal/app/controllers/developer_portal/login_controller.rb

```diff
@@ -26,14 +27,16 @@ def new
   def create
     logout_keeping_session!
 
-    if (@user = @strategy.authenticate(params.merge(request: request)))
+    if sign_in_user
       self.current_user = @user
       create_user_session!
       flash[:notice] = @strategy.new_user_created? ? 'Signed up successfully' : 'Signed in successfully'
       redirect_back_or_default(@strategy.redirect_to_on_successful_login)
     elsif @strategy.redirects_to_signup?
       @strategy.on_signup(session)
       redirect_to @strategy.signup_path(params), notice: 'Successfully authenticated, please complete the signup form'
+    elsif !spam_check(@user)
+      render_creation_error('Spam protection failed.')
```

Comment:
I can't understand how is this supposed to work. Under `sign_in_user` we have
```
 @user = @strategy.authenticate(params.merge(request: request))
```

So if that returns a user, then we don't go to spam protection. If it returns `nil` or `false`, then what is the point in passing `@user` as a parameter to `spam_check(@user)`?

I assume I'm missing details how this works. So excuse my ignorance.

Reply:
You're not missing any details. I, on the contrary, am taking shots in the dark since I don't understand very well how the whole spam check flow works.
Ideally, `@user` is set inside `sign_in_user` so that it can be passed to `spam_check`. I am trying to replicate the same code from reset_password.

Reply:
@akostadinov is right. If `#sign_in_user` sets a `@user`, you'll never get to `elsif !spam_check(@user)`. On the other hand, if it does not set a `@user`, then `@user` will be falsy and I don't see how `#spam_check` can decide based on that (`nil` or `false`? not cool, so no).

Reply:
Fixed with my commit